osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/mbus/IdentificationNumberFactory.java

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright 2019 Smart Society Services B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.mbus;
+
+import org.apache.commons.lang3.StringUtils;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.CommandExecutorDeviceContext;
+
+/**
+ * Handles creation and validation of @{@link IdentificationNumber} based on either the String representation or the
+ * Number representation.
+ */
+class IdentificationNumberFactory {
+
+    private static final int HEX_RADIX = 16;
+    private static final String IDENTIFICATION_NUMBER_REGEX_DSMR4 = "\\d{1,8}";
+    private static final String IDENTIFICATION_NUMBER_REGEX_SMR5 = "[0-9A-Fa-f]{1,8}";
+    private static final long MIN_IDENTIFICATION = 0;
+    private static final long MAX_IDENTIFICATION = Long.parseLong("99999999", HEX_RADIX);
+
+    private final CommandExecutorDeviceContext deviceContext;
+
+    private IdentificationNumberFactory(final CommandExecutorDeviceContext deviceContext) {
+        this.deviceContext = deviceContext;
+    }
+
+    public static IdentificationNumberFactory create(CommandExecutorDeviceContext deviceContext) {
+        return new IdentificationNumberFactory(deviceContext);
+    }
```

Comment:
This is quite a neat pattern, creating a device-specific factory to encapsulate protocol specific logic. However there are some drawbacks:

- the whole DlmsDevice is passed in, but only the protocol (basically DSMR4 vs DMR5) is needed. Too much information is shared
- the logic needs to use conditions throughout to handle the different cases depending on the protocol

So I think we can do better than this. I suggest creating separate classes for the DSMR4 / SMR5 specific logic with an abstract super class which applies the Template Method pattern. See for an example (created by @stefanermens after a similar review comment by me on a PR of his) `org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.dlmsobjectconfig.DlmsObjectConfig`

There needs to be a factory which determines which implementation to use based on the protocol and then the drawbacks mentioned above are avoided.

Reply:
indeed, if the Context object will be removed, the scope can be narrowed.

As for the condition, there is only one, so it seems a bit overkill to make separate classes for only the regex.